File: icpc/uw23/B.cpp
// s=1, t=2
// - F S A -> s=(1,0), t=(0,0)
//   - P S T -> s=(0,0), t=(1,0)
//     - F T A -> s=(0,0), t=(2,0)
//     - F T B -> s=(0,0), t=(1,1)
//     - F S A -> s=(1,0), t=(1,0)

/* achieve 1/2:
 * Fill S with A, Pour S into T, Fill T with B -> t=(1, 1)
 * Pour T into S, Empty T, Pour S into T -> t=(1/2, 1/2)
 * Fill T with B -> t=(1/2, 3/2)
 * Pour T into S, Empty T, Pour S into T -> t=(1/4, 3/4)
 * Fill T with A -> t=(5/4, 3/4)
 * Pout T into S, Empty T, Pour S into T -> t=(5/8, 3/8)
 * Fill T with B -> t=(5/8, 11/8)
 * 
 * From (a/2^k, (2^k+1 - a)/2^k) in T, we can get ((a + 2^k+1)/2^k+1, (2^k+1 - a)/2^k+1) or (a/2^k+1, (2^k+2 - a)/2^k+1) in T
 */

// max s = 1, max t = 2
// s = (A_S, B_S), t = (A_T, B_T)
// * F S A : s -> (1 - B_S, B_S)
// * F S B : s -> (A_S, 1 - A_S)
// * F T A : t -> (2 - B_T, B_T)
// * F T B : t -> (A_T, 2 - A_T)
// * P S T : l = min(1, 2 - A_T - B_T), t -> (l * A_S + A_T, l * B_S + B_T), s -> (A_S * (1 - l), B_S * (1 - l))
// * 

File: icpc/uw23/B.py
from dataclasses import dataclass
import sys
from math import gcd

@dataclass
class Frac:
    n: int
    d: int

    def whole(N: int):
        return Frac(N, 1)

    def zero():
        return Frac.whole(0)

    def __add__(self, other):
        n = self.n * other.d + other.n * self.d
        d = self.d * other.d
        g = gcd(n, d)
        return Frac(n // g, d // g)
    
    def __sub__(self, other):
        n = self.n * other.d - other.n * self.d
        d = self.d * other.d
        g = gcd(n, d)
        return Frac(n // g, d // g)

    def __mul__(self, other):
        n = self.n * other.n
        d = self.d * other.d
        g = gcd(n, d)
        return Frac(n // g, d // g)
    
    def __truediv__(self, other):
        n = self.n * other.d
        d = self.d * other.n
        g = gcd(n, d)
        return Frac(n // g, d // g)
    
    def __lt__(self, other):
        return self.n * other.d < other.n * self.d


@dataclass
class Beaker:
    cap: Frac
    A: Frac = Frac.zero()
    B: Frac = Frac.zero()


a, b, s, t = map(int, sys.stdin.readline().split())

S = Beaker(Frac.whole(s))
T = Beaker(Frac.whole(t))
log = []

def fill(x: str, y: str):
    X = S if x == "S" else T
    if y == "A":
        X.A = X.cap - X.B
    else:
        X.B = X.cap - X.A
    log.append(f"F {x} {y}")

def pour(x: str, y: str):
    X, Y = (S, T) if x == "S" else (T, S)
    rem = Y.cap - (Y.A + Y.B)
    if rem < (X.A + X.B):
        t = rem / (X.A + X.B)
        Y.A += t * X.A
        Y.B += t * X.B
        X.A -= t * X.A
        X.B -= t * X.B
    else:
        Y.A += X.A
        Y.B += X.B
        X.A = Frac.zero()
        X.B = Frac.zero()

    log.append(f"P {x} {y}")

def empty(x: str):
    X = S if x == "S" else T
    X.A = Frac.zero()
    X.B = Frac.zero()
    log.append(f"E {x}")

def seal(x: str):
    log.append(f"! {x}")

def answer():
    print(len(log))
    print("\n".join(log))

fill("T", "A")
pour("T", "S")
empty("S")
fill("T", "B")
pour("T", "S")
empty("T")
pour("S", "T")
fill("T", "A")
seal("T")
print(S, T)
answer()